server/etcdserver/raft.go

```diff
@@ -314,6 +327,43 @@ func (r *raftNode) start(rh *raftReadyHandler) {
 	}()
 }
 
+// For a cluster with only one member, the raft always send identical
+// unstable entries and committed entries to etcdserver, and etcd
+// responds to the client once it finishes (actually partially) the
+// applying workflow.
+//
+// When the client receives the response, it doesn't mean etcd has already
+// successfully saved the data, including BoltDB and WAL, because:
+//    1. etcd commits the boltDB transaction periodically instead of on each request;
+//    2. etcd saves WAL entries in parallel with applying the committed entries.
+// Accordingly, it might run into a situation of data loss when the etcd crashes
+// immediately after responding to the client and before the boltDB and WAL
+// successfully save the data to disk.
+// Note that this issue can only happen for clusters with only one member.
+//
+// For clusters with multiple members, it isn't an issue, because etcd will
+// not commit & apply the data before it being replicated to majority members.
+// When the client receives the response, it means the data must have been applied.
+// It further means the data must have been committed.
+// Note: for clusters with multiple members, the raft will never send identical
+// unstable entries and committed entries to etcdserver.
+func shouldWaitWALSync(unstableEntries []raftpb.Entry, committedEntries []raftpb.Entry) bool {
+	if len(unstableEntries) != len(committedEntries) {
+		return false
+	}
+
+	if len(unstableEntries) == 0 {
+		return false
+	}
+
+	// Previously `reflect.DeepEqual` is used to compare the `unstableEntries`
+	// and `committedEntries`, but it's a little expensive, so we just compare
+	// the log index. The applying workflow should wait for the WAL sync if
+	// there is any log index overlap between them.
+	lastCommittedEntry := committedEntries[len(committedEntries)-1]
+	return lastCommittedEntry.Term >= unstableEntries[0].Term && lastCommittedEntry.Index >= unstableEntries[0].Index
```

Comment:
```suggestion
	return lastCommittedEntry.Term > unstableEntries[0].Term || (lastCommittedEntry.Term == unstableEntries[0].Term && lastCommittedEntry.Index >= unstableEntries[0].Index)
```

Reply:
I thought about this before, and the comment is only partially correct. Specially, it isn't correct in this case. Points:
1. The index indeed `might` go down, because unstable(uncommitted) log might be overwritten by new leader in a new term. But committed log will never be overwritten. It means the index will never go down to the committed index, in other words, the index will always be bigger than the committed index. 
2. Especially for a cluster with only one member, there is no leader change at all. So this comment doesn't stand. 

Both the original code and your suggest code work, but your suggested change has better readability, so updated per your suggestion.